File: src/main/java/io/github/phantamanta44/libnine/util/world/structmatcher/StructureMatcherBlockState.java
package io.github.phantamanta44.libnine.util.world.structmatcher;

import io.github.phantamanta44.libnine.util.world.WorldBlockPos;
import net.minecraft.block.state.IBlockState;
import net.minecraft.util.math.Vec3i;

import java.util.List;
import java.util.function.Predicate;

public class StructureMatcherBlockState implements IStructureMatcher {

    private final Predicate<IBlockState> condition;

    public StructureMatcherBlockState(Predicate<IBlockState> condition) {
        this.condition = condition;
    }

    @Override
    public boolean testStructure(WorldBlockPos basePos, List<Vec3i> components) {
        for (Vec3i pos : components) {
            if (!condition.test(basePos.offset(pos).getBlockState())) {
                return false;
            }
        }
        return true;
    }

}

File: src/main/kotlin/xyz/phanta/libnine/util/component/multiblock/matcher/StructureMatcherBlockState.kt
package xyz.phanta.libnine.util.component.multiblock.matcher

import net.minecraft.block.BlockState
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Vec3i
import net.minecraft.world.World
import xyz.phanta.libnine.util.world.plus

class StructureMatcherBlockState(private val condition: (BlockState) -> Boolean) : StructureMatcher {

    override fun testStructure(world: World, basePos: BlockPos, components: List<Vec3i>): Boolean =
            components.all { condition(world.getBlockState(basePos + it)) }

}
